### PR TITLE
Generalised to multiple runtime directories with priorities

### LINKED ISSUE #3346: Treat runtime files from installation as immutable
We should treat the grammar sources, compiled parsers, and queries produced while building Helix as immutable defaults which can be overridden with the existence of files in `~/.config/helix/runtime`. Effectively, `$HELIX_RUNTIME` should act like `$PATH` in unix.

Except when installing from source, these files almost always end up in some `/usr` directory that needs sudo to edit (which you probably shouldn't do). With Nix, these end up in the store which is immutable so you need to build from source to modify queries or themes locally. Instead you should just be able to override files in your `~/.config/helix/runtime` directory.

See https://github.com/helix-editor/helix/issues/3310#issuecomment-1203875141
See https://github.com/helix-editor/helix/issues/2091#issuecomment-1097472407

### PR DESCRIPTION
**Pull request by paul-scott**
This is an implementation for #3346 
See commit message for an overview.

Here are some thing to consider when reviewing.

Depending on how helix is packaged, it still might not be possible for some users to conveniently override runtime files. E.g. on AUR helix-git, `/usr/bin/hx` is a shell script that sets `HELIX_RUNTIME` to `/usr/lib/helix/runtime` before calling `/usr/lib/helix/hx`. Setting this env var is not necessary because helix already automatically checks for a runtime directory in the same directory as the executable. Setting it prevents `~/.config/helix/runtime` from getting priority from the default runtime installation.

[The docs suggest](https://docs.helix-editor.com/master/install.html#build-from-source) that the `HELIX_RUNTIME` is intended as an option to override a `~/.config/helix/runtime`, which I agree could be useful and why I've initially kept this behaviour. However, if it looks like packager's and build-from-source installers want to use this env var for their default runtime install, then we might want to rethink the priority order as outlined in the commit message.

For consistencies sake you might want to consider if `~/.config/helix/themes` should be deprecated since this commit makes it so that `~/.config/helix/runtime/themes` can be used to the same effect. Before and after this commit the former always has the higher priority.

It just remembered that I haven't touched the user doc / book. I'll have to have a look and see where the new behaviour can be easily explained.

Closes #3346

### CODE REVIEW

File: helix-loader/src/lib.rs
Comment (archseer) [inline helix-loader/src/lib.rs:82]: I'd avoid `AsRef`/monomorphization here since it can grow the binary by generating a lot of alternative method bodies. Just take `&Path` here

File: helix-loader/src/lib.rs
Comment (archseer) [inline helix-loader/src/lib.rs:39]: It seems to me that this could be done in a `Lazy` block and stored as `static RUNTIME_DIRS` so that we only compute this vector once

Comment (archseer) [reply helix-loader/src/lib.rs:39]: Ah sorry, I missed that this was actually done above!

**Comment by archseer** [comment]
Priority should be (highest to lowest):

1. `~/.config/helix/runtime` (user's configuration)
2. `HELIX_RUNTIME` (indicates the system's runtime, will be usually be under `/usr/` for package installs, not writable)
3. current executable's directory (if all else fails, it might just be someone running it from a tarball)

This means that for grammar fetch & build we probably want to write into the first dir? Except on CI \cc @the-mikedavis

File: helix-term/src/health.rs
Comment (archseer) [inline helix-term/src/health.rs:81]: I'd just output this similar to `PATH` or `GOPATH`: a bunch of directories separated by `;`

**Comment by paul-scott** [comment]
> Priority should be (highest to lowest):
> 
>     1. `~/.config/helix/runtime` (user's configuration)
> 
>     2. `HELIX_RUNTIME` (indicates the system's runtime, will be usually be under `/usr/` for package installs, not writable)
> 
>     3. current executable's directory (if all else fails, it might just be someone running it from a tarball)
> 
> 
> The docs recommend setting `HELIX_RUNTIME` just because I felt it's simpler to use an env var than to copy files over and then forget to update them in the future.
> 
> This means that for grammar fetch & build we probably want to write into the first dir? Except on CI \cc @the-mikedavis

In that case how about:

1. sibling `runtime` directory to `$CARGO_MANIFEST_DIR`
2. subdirectory of user config directory (`~/.config/helix/runtime`)
3. `$HELIX_RUNTIME`
4. subdirectory of path to helix executable

Compared to the original behaviour this is just swapping 2 and 3.

I assume my 1 here is what the CI uses. In addition it seems like `helix-git` on AUR  fetches and builds the grammars under 1, and then packages them up from there so they get installed under `/usr/lib/helix/runtime`.

I think it is best if this kind of packaging behaviour remains. It would be a pain if on updating a packager started putting (and overriding) the grammars under `~/.config/helix/runtime`. A user might intentionally be trying to keep grammars in there to override the defaults.

**Comment by archseer** [comment]
Oh I agree, I forgot to list it above :) Just wanted to clarify the intended use of `HELIX_RUNTIME`

File: helix-term/src/health.rs
Comment (paul-scott) [reply helix-term/src/health.rs:81]: In addition to the runtime paths, the original behaviour conditionally reports for each runtime directory:

1. the destination if a runtime directory is symlinked
2. a red warning if the runtime directory doesn't exist
3. a red warning if the runtime directory is empty

I assume you still want this info displayed? How should we clearly display this using a `;`-separated list?  For 2 and 3, we can just filter on the non-existent / empty directories and join them on `;`. Should we just do the same thing for 1, and leave it up to the user to work out *which* runtime directory is the symlink?

**Comment by gibbz00** [comment]
Should we also make use of XDG_RUNTIME_DIR? I'm personally not of fan of placing runtime files the .config folder.

**Comment by archseer** [comment]
It's orthogonal to this PR. https://github.com/helix-editor/helix/pull/2135 implements that but it should be rebased after this is merged

File: helix-term/src/health.rs
Comment (paul-scott) [reply helix-term/src/health.rs]: I've change the implementation so that a `;`-separated list of runtime directories is presented. For the warnings / errors,  changed them all to warnings (yellow) because now it might be more common for a runtime directory to be missing or empty and it not be a problem. I didn't join these paths on `;` in the warnings, instead I just report a warning per case.

One thing is that windows / linux use `;` / `:` for this separator in `PATH` and `GOPATH`. I'm not sure we want to go to the effort to make this OS aware, so I just went with your suggestion for the semicolon. This seems like it would cause the least confusion given windows using `:` after drive letter in path `C:\\` etc.

Let me know what you think.

**Comment by paul-scott** [comment]
> Oh I agree, I forgot to list it above :) Just wanted to clarify the intended use of `HELIX_RUNTIME`

I've implemented pushing `HELIX_RUNTIME` down in priority below the user config runtime.

In terms of fetch and build, the behaviour is now that a cargo build will always put grammars in the runtime directory within the helix repo being built. `HELIX_RUNTIME` can no longer be used to change where they will automatically be fetched / built.

When a user calls `hx -g fetch / build` this is now always put in `~/.config/helix/runtime/grammars/`.

None of this is hard-coded, it is just the way it works with the new priority ordering and how the implementation of fetch and build always grab the highest priority runtime directory.

I've modified the build from source instructions in the book to reflect some of this but not in that much detail.

Let me know if you would like to further tweak this.

File: helix-view/src/theme.rs
Comment (gibbz00) [inline helix-view/src/theme.rs:61]: Hi, I'm trying to merge this into #5636. What's the point of this `HashSet` introduction?

Comment (gibbz00) [reply helix-view/src/theme.rs:61]: Nevermind, sorry. Didn't see the new `load_theme` logic.

File: helix-view/src/theme.rs
Comment (pascalkuthe) [inline helix-view/src/theme.rs:175]: I think we should track whether there is a cycle here (simply set a bool to true if we return `None` because `visited_paths` already contains `path` so we can create a separate error message for that case. Having the same error message for both cases seems confusing.

**Review by pascalkuthe — CHANGES_REQUESTED** [review]
This looks great! Just some minor UI nits

File: book/src/install.md
Comment (paul-scott) [reply book/src/install.md]: Done. I've also moved it down slightly in the document to try and make the flow make a bit more sense.

File: helix-term/src/health.rs
Comment (paul-scott) [reply helix-term/src/health.rs]: Have added both paths to output.

File: helix-view/src/theme.rs
Comment (paul-scott) [reply helix-view/src/theme.rs]: This method is actually no longer used anywhere . I'm not sure if this is due to change in master being merged or if I made the change. Anyway I've removed the method.

File: helix-view/src/theme.rs
Comment (paul-scott) [reply helix-view/src/theme.rs]: Yep I'm doing that now. This method now returns a Result which distinguishes between the two cases.

File: book/src/install.md
Comment (paul-scott) [reply book/src/install.md]: There was a conflict with master. I've now merged in latest master and resolved the conflict.

BTW, what is the preferred process to keep these PRs up to date / conflict free with master? Should we be periodically rebasing and force pushing to our PR branch, or doing a regular merge? I was doing the former to try and keep things cleaner but then I see archseer at some point decided to do a regular merge.

Comment (archseer) [reply book/src/install.md]: Rebase is preferred but if it's a small PR that we'll squash then it'll result in a single commit anyway

File: helix-term/src/ui/mod.rs
Comment (paul-scott) [reply helix-term/src/ui/mod.rs:284]: Look down a couple of lines to 289 and 290 and it looks like this being done already.

File: book/src/install.md
Comment (paul-scott) [reply book/src/install.md:200]: Sounds good I'll make this change.

File: helix-term/src/ui/mod.rs
Comment (pascalkuthe) [reply helix-term/src/ui/mod.rs:284]: Ah you are right, sorry about that 😅

**Review by pascalkuthe — APPROVED** [review]
LGTM now

**Comment by archseer** [comment]
Thanks! This was a much needed improvement 🎉

File: book/src/install.md
Comment (flokli) [inline book/src/install.md:202]: Hmm, shouldn't this be  ~/.local/state/helix/runtime (`XDG_STATE_HOME`)? 

This is clearly not config, but state, and we don't want backup programs to back up compiled tree-sitter .so grammars…

Comment (the-mikedavis) [reply book/src/install.md:202]: That's https://github.com/helix-editor/helix/issues/584

This PR kept it to `~/.config` for backwards compatibility

Comment (gibbz00) [reply book/src/install.md:202]: https://github.com/helix-editor/helix/pull/5636 🤷

**Pull request merged**
